pkg/clustermanager/kube_proxy.go

```diff
@@ -0,0 +1,505 @@
+		opt(u)
+	}
+
+	return *u
+}
+
+// KubeProxyCLIUpgrader prepares a cluster for a kube-proxy upgrade.
+// It's mostly a wrapper around [KubeProxyUpgrader] to be used from the CLI.
+// It builds clients from kubeconfig files and facilitates mocking. It also uses a retrier
+// around [KubeProxyCLIUpgrader] to deal with transient errors.
+type KubeProxyCLIUpgrader struct {
+	clientFactory ClientFactory
+	log           logr.Logger
+	retrier       retrier.Retrier
+}
+
+// KubeProxyCLIUpgraderOpt allows to customize a KubeProxyCLIUpgraderOpt
+// on construction.
+type KubeProxyCLIUpgraderOpt func(*KubeProxyCLIUpgrader)
+
+// KubeProxyCLIUpgraderRetrier allows to use a custom retrier.
+func KubeProxyCLIUpgraderRetrier(retrier retrier.Retrier) KubeProxyCLIUpgraderOpt {
+	return func(u *KubeProxyCLIUpgrader) {
+		u.retrier = retrier
+	}
+}
+
+// PrepareUpgrade perfoms the necessary steps prior to a kube-proxy upgrade.
+func (u KubeProxyCLIUpgrader) PrepareUpgrade(ctx context.Context,
+	spec *cluster.Spec,
+	managementClusterKubeconfigPath, workloadClusterKubeconfigPath string,
+) error {
+	managementClusterClient, workloadClusterClient, err := u.buildClients(
+		managementClusterKubeconfigPath, workloadClusterKubeconfigPath,
+	)
+	if err != nil {
+		return err
+	}
+
+	up := NewKubeProxyUpgrader()
+
+	return u.retrier.Retry(func() error {
+		return up.PrepareForUpgrade(ctx, u.log, managementClusterClient, workloadClusterClient, spec)
+	})
+}
+
+// CleanupAfterUpgrade perfoms the necessary steps after an upgrade.
+func (u KubeProxyCLIUpgrader) CleanupAfterUpgrade(ctx context.Context,
+	spec *cluster.Spec,
+	managementClusterKubeconfigPath, workloadClusterKubeconfigPath string,
+) error {
+	managementClusterClient, workloadClusterClient, err := u.buildClients(
+		managementClusterKubeconfigPath, workloadClusterKubeconfigPath,
+	)
+	if err != nil {
+		return err
+	}
+
+	up := NewKubeProxyUpgrader()
+
+	return u.retrier.Retry(func() error {
+		return up.CleanupAfterUpgrade(ctx, u.log, managementClusterClient, workloadClusterClient, spec)
+	})
+}
+
+func (u KubeProxyCLIUpgrader) buildClients(
+	managementClusterKubeconfigPath, workloadClusterKubeconfigPath string,
+) (managementClusterClient, workloadClusterClient client.Client, err error) {
+	u.log.V(4).Info("Building client for management cluster", "kubeconfig", managementClusterKubeconfigPath)
+	managementClusterClient, err = u.clientFactory.BuildClientFromKubeconfig(managementClusterKubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u.log.V(4).Info("Building client for workload cluster", "kubeconfig", workloadClusterKubeconfigPath)
+	workloadClusterClient, err = u.clientFactory.BuildClientFromKubeconfig(workloadClusterKubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return managementClusterClient, workloadClusterClient, nil
+}
+
+// NewKubeProxyUpgrader builds a new KubeProxyUpgrader.
+func NewKubeProxyUpgrader(opts ...KubeProxyUpgraderOpt) KubeProxyUpgrader {
```

Comment:
I don't think it will need to be used in the controller. I think we can remove this on the next release.

If we decide not to, the logic to implement this in the controller will be similar, but not the same. For example, we won't use a retrier in the "ensureUpdate" method.

The reason it's two objects is it because it helps separate concerns (business logic from retrying logic). Originally the non CLI one was just a standalone function, but I made it a method to be able to introduce the "timing" options, only because of testing.